Lib/test/test_iter.py

```diff
@@ -238,6 +244,52 @@ def test_mutating_seq_class_exhausted_iter(self):
         self.assertEqual(list(empit), [5, 6])
         self.assertEqual(list(a), [0, 1, 2, 3, 4, 5, 6])
 
+    def test_reduce_mutating_builtins_iter(self):
+        # This is a reproducer of issue #101765
+        # where iter `__reduce__` calls could lead to a segfault or SystemError
+        # depending on the order of C argument evaluation, which is undefined
+
+        # Backup builtins.iter
+        builtins = __builtins__.__dict__ if hasattr(__builtins__, "__dict__") else __builtins__
+        orig_iter = builtins["iter"]
+
+        def run(item):
+            (fn, *flag), *initializer = item
+            it = iter(fn(*initializer), *flag)
+
+            class CustomStr:
+                def __hash__(self):
+                    return hash("iter")
+                def __eq__(self, other):
+                    # Here we exhaust `it`, possibly changing our `it_seq` pointer to NULL
+                    # The `__reduce__` call should correctly get the pointers after this call
+                    list(it)
+                    return other == "iter"
+
+            del builtins["iter"]
```

Comment:
It might be worth a comment here clarifying that we have to first `del` because otherwise the next line won't replace the key in the dict; otherwise it would be easy for someone to unthinkingly remove this `del` as a no-op

Reply:
Added some additional comments for this and other test elements in [178b8ea](https://github.com/python/cpython/pull/101769/commits/178b8ea41bb045eed63d06674e2f337d4de97bef)